File: 75-sort-colors/sort-colors.java
class Solution {

    public static void swap(int[] nums, int first, int second) {
        int temp = nums[first];
        nums[first] = nums[second];
        nums[second] = temp;
    }

    public void sortColors(int[] nums) {
        int redMarker = 0;
        int blueMarker = nums.length - 1;
        int marker = 0;

        while(marker <= blueMarker) {
            if(nums[marker] == 0) {
                swap(nums, marker, redMarker);
                redMarker++;
                marker = redMarker;
            } else if(nums[marker] == 2) {
                swap(nums, marker, blueMarker);
                blueMarker--;
            } else {
                marker++; 
            }
        }
    }
}

File: 75-sort-colors/sort-colors.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        RED = 0
        WHITE = 1
        BLUE = 2

        # partition such that red goes in the front and blue marker goes in teh back
        # 

        curr = 0
        left = 0
        right = len(nums) - 1

        while curr <= right and left < right:
            if nums[curr] == RED:
                nums[curr], nums[left] = nums[left], nums[curr]
                left += 1
                curr = left
            elif nums[curr] == BLUE:
                nums[curr], nums[right] = nums[right], nums[curr]
                right -= 1
            else:
                curr += 1
        


        